tests/callbacks/test_quantization.py

```diff
@@ -152,3 +162,139 @@ def test_quantization_triggers(tmpdir, trigger_fn: Union[None, int, Callable], e
+                # ``torch.quantization.FakeQuantize`` checks observer_enabled[0] == 1.
+                observer_enabled = observer_enabled[0] == 1
+                assert observer_enabled == expected_observer_enabled
+            else:
+                assert torch.equal(observer_enabled, expected_observer_enabled)
+
+    def _check_observer_state(self, should_disable):
+        if should_disable:
+            self._assert_observer_enabled(collections.defaultdict(lambda: False))
+        else:
+            self._assert_observer_enabled(self._last_train_fake_quant_to_observer_enabled)
+
+    def on_fit_start(self):
+        self._fake_quants = tuple(
+            module for module in self.modules() if isinstance(module, torch.quantization.FakeQuantizeBase)
+        )
+        self._fake_quant_to_initial_state = {fake_quant: fake_quant.state_dict() for fake_quant in self._fake_quants}
+        self._last_train_fake_quant_to_observer_enabled = self._collect_observer_enabled()
+
+    def training_step(self, batch, batch_idx):
+        self._assert_observer_enabled(self._last_train_fake_quant_to_observer_enabled)
+
+        result = super().training_step(batch, batch_idx)
+        if self._disable_observers_after_step == self.global_step:
+            self.apply(torch.quantization.disable_observer)
+
+        self._last_train_fake_quant_to_observer_enabled = self._collect_observer_enabled()
+
+        return result
+
+    def on_validation_start(self):
+        self._check_observer_state('validate' in self._disable_observers and not self.trainer.sanity_checking)
+
+    def on_validation_end(self):
+        if 'validate' in self._disable_observers:
+            if self.trainer.sanity_checking:
+                self._assert_fake_quant_initial_state()
+            else:
+                self._assert_observer_enabled(self._last_train_fake_quant_to_observer_enabled)
+
+    def on_test_start(self):
+        self._check_observer_state('test' in self._disable_observers)
+
+    def on_test_end(self):
+        if 'test' in self._disable_observers:
+            self._assert_observer_enabled(self._last_train_fake_quant_to_observer_enabled)
+
+    def on_predict_start(self):
+        self._check_observer_state('predict' in self._disable_observers)
+
+    def on_predict_end(self):
+        if 'predict' in self._disable_observers:
+            self._assert_observer_enabled(self._last_train_fake_quant_to_observer_enabled)
+
+
+@pytest.mark.parametrize("observe", ['average', 'histogram'])
+@pytest.mark.parametrize("disable_observers", [('validate', 'test', 'predict'), ('validate', ), ()])
+@RunIf(quantization=True)
+def test_disable_observers(tmpdir, observe, disable_observers):
+    """Test disabling observers"""
+    num_features = 16
+    dm = RegressDataModule(num_features=num_features, length=800, batch_size=10)
+    qmodel = CheckObserverDisabledModel(disable_observers, disable_observers_after_step=2)
```

Comment:
Should `disable_observers_after_n_step` be a parameter of the Callback ?

Reply:
The observers could be disabled or enabled in the middle of training. For example, users may

- disable observers after 100 steps.
- disable observers after 2 epochs.
- disable an observer if the quantization range is too large.
...

The test check if the callback properly handle the above cases. It seems to me that `disable_observers_after_n_step` is API improvement, should we add `disable_observers_after_n_step` to the `QuantizationAwareTraining` callback  in this PR?

Reply:
Yes, I think it could be a nice API improvement. Maybe another PR ?